charger/fritzdect.go

```diff
@@ -0,0 +1,333 @@
+		return fmt.Errorf("wasn't able to switchOff: %d", status.State)
+	}
+
+	return nil
+}
+
+// MaxCurrent implements the Charger.MaxCurrent interface (Dummy function)
+func (c *FritzDECT) MaxCurrent(current int64) error {
+	// FritzDECT switch has no option to set MaxCurrent
+	return nil
+}
+
+// CurrentPower implements the Meter interface.
+func (c *FritzDECT) CurrentPower() (float64, error) {
+	status, err := c.apiStatus()
+	if status.Power < c.standbypower {
+		return 0, err
+	}
+	return status.Power, err
+}
+
+// ChargedEnergy implements the ChargeRater interface
+func (c *FritzDECT) ChargedEnergy() (float64, error) {
+	status, err := c.apiStatus()
+
+	energy := status.Energy / 1000
+
+	return energy, err
+}
+
+// Fritzbox helpers (thx to ahago)
+
+//getFritzBoxSessionID fetches a session-id based on the username and password in the connection struct
+func (c *FritzDECT) getFritzBoxSessionID() error {
+	uri := fmt.Sprintf("%s/login_sid.lua", c.uri)
+	parameters := make(map[string]string)
+	type Result struct {
+		SID       string
+		Challenge string
+		BlockTime string
+	}
+
+	body, err := sendFritzBoxRequest(uri, parameters)
+	if err != nil {
+		return err
+	}
+
+	v := Result{SID: "none", Challenge: "none", BlockTime: "none"}
+	err = xml.Unmarshal(body, &v)
+	if err != nil {
+		return err
+	}
+
+	if v.SID == "0000000000000000" {
+		parameters["username"] = c.user
+		parameters["response"] = createFbChallengeResponse(v.Challenge, c.password)
+		body, err = sendFritzBoxRequest(uri, parameters)
+		if err != nil {
+			return err
+		}
+		err = xml.Unmarshal(body, &v)
+		if err != nil {
+			return err
+		}
+	}
+	c.sid = v.SID
+	return nil
+}
+
+// sendFritzBoxRequest sends a HTTP Request to the FritzBox based on the given URL and returns the answer as a bytearray
+func sendFritzBoxRequest(baseURL string, parametersIn map[string]string) ([]byte, error) {
+	var URL *url.URL
+	URL, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	parameters := url.Values{}
+	for key, value := range parametersIn {
+		parameters.Add(key, value)
+	}
+	URL.RawQuery = parameters.Encode()
+
+	resp, err := http.Get(URL.String())
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, errors.New("sendFritzBoxRequest status code != 200")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return body, nil
+}
+
+// createFbChallengeResponse creates the Fritzbox challenge response string
+func createFbChallengeResponse(challenge string, pass string) string {
+	// Create Response string to be crypted
+	utf8 := []byte(challenge + "-" + pass)
+	utf16le := encFbUTF8ToUTF16le(utf8)
+	hash := md5.New()
+	hash.Write(utf16le)
+	md5hash := hex.EncodeToString(hash.Sum(nil))
+	return challenge + "-" + md5hash
+}
+
+// encFbUTF8ToUTF16le encodes an UTF8 byte array to an UTF16LE byte array
+func encFbUTF8ToUTF16le(in []byte) []byte {
```

Comment:
https://forum.golangbridge.org/t/how-to-convert-utf-8-string-to-utf-16-be-string/7072?